danlp/models/bert_models.py

```diff
@@ -201,68 +220,93 @@ def __init__(self, cache_dir=DEFAULT_CACHE_DIR, verbose=False):
         self.max_length_sub = self.model_sub.bert.embeddings.position_embeddings.num_embeddings
         self.max_length_pol = self.model_pol.bert.embeddings.position_embeddings.num_embeddings
 
-        
+        if device is not None:
+            self.device = torch.device(device)
+            self.model_sub.to(self.device)
+            self.model_pol.to(self.device)
+        else: self.device = device
+        self.parallel = parallel
+        if self.parallel:
+            self.model_sub = torch.nn.DataParallel(self.model_sub)
+            self.model_pos = torch.nn.DataParallel(self.model_pos)
+
     def _clean(self, sentence):
-        sentence=re.sub(r'http[^\s]+', '', sentence)
-        sentence = re.sub(r'\n', '',sentence)
-        sentence = re.sub(r'\t', '',sentence)
-        sentence = re.sub(r'@', '',sentence)
-        sentence = re.sub(r'#', '',sentence)
+        sentence = re.sub(r'http[^\s]+', '', sentence)
+        sentence = re.sub(r'\n', '', sentence)
+        sentence = re.sub(r'\t', '', sentence)
+        sentence = re.sub(r'@', '', sentence)
+        sentence = re.sub(r'#', '', sentence)
         return sentence
 
     def _classes(self):
         return self.classes_pol, self.classes_sub
-    
-    def _get_pred(self, tokenizer, model, max_lenght, sentence):
+
+    def _get_pred(self, tokenizer, model, max_length, sentence):
+        if isinstance(sentence, str):  # to ensure common pipeline
+            sentence = [sentence]
         input1 = tokenizer.encode_plus(sentence, add_special_tokens=True, return_tensors='pt',
-                                                max_length=max_lenght, return_overflowing_tokens=True)
+                                       max_length=max_length, return_overflowing_tokens=True,
+                                       truncation=True,
+                                       padding=True)
         if 'overflowing_tokens' in input1:
-            warnings.warn('Maximum length for sequence exceeded, truncation may result in unexpected results. Consider running the model on a shorter sequenze then {} tokens'.format(max_lenght))
-        pred = model(input1['input_ids'], token_type_ids=input1['token_type_ids'])[0]
+            warnings.warn(
+                'Maximum length for sequence exceeded, truncation may result in unexpected results. Consider running the model on a shorter sequenze then {} tokens'.format(max_length))
+        if self.device:
+            input1.to(self.device)  # move files to GPU if specified
+        else:
+            pred = model(**input1)[0]
 
         return pred
```

Comment:
From the previous comment - it seems to return the same value for pred every time. 

@AmaliePauli have you checked that the tokenizer tokenizes it correctly? The input to the function itself follows the guide by huggingface and should be valid.

Reply:
Hi @KennethEnevoldsen, sorry for the response time!

The models and the code in the Danlp are using an older version of the transformers==2.3.0. I was wondering if you are using a newer version in the documentation you are reffering to? 
Because, I got several issues when running the code; 
The output is the same for every sentence, and yes it might look like the error happens whne providing a list to the tokenizer - so pehaps it is a verison issue, if you are not getting the same output for every sentence.
Also, providing a list of sentence as input  in predict() results only in "one output" like, 
classifier.predict( ['Analysen viser, at økonomien bliver forfærdelig dårlig', 'Jeg tror alligvel, det bliver godt'])
   {'analytic': 'objective', 'polarity': 'neutral'}

I did tried the code on GPU, and got the following error in _get_pred:
    input1.to(self.device)  # move files to GPU if specified
AttributeError: 'dict' object has no attribute 'to'

At some time we will proberly update to a newer transformer version and then we also want to retrain the models. 

Kind regards
Amalie

Reply:
yep probably I am using 3.1.0, which I believe is the newest version at the time of writing?


```
input1.to(self.device) # move files to GPU if specified
AttributeError: 'dict' object has no attribute 'to'
```
This is interesting I have had a colleague with the same issue so versioning does seem to be a problem. It is because the newer version returns a batchEncoding class instead of a dictionary class.

(is there an issue with updating the version without retraining the models? At least the models loaded in just fine it 3.1.0)

Reply:
I tried with transformer 3.1.0, however, an error occurs in using get_pred() on gpu because "pred" is never assigned if device is not None. And the other issues running on the CPU still applies. 

Perhaps not, so if all the benchmarks is the same with the newer version, yes then there will be no need for retraining.

Reply:
hmm that indeed seems to be the case. I have changed it now